components/base/kinematicbase/ptgKinematics.go

```diff
@@ -211,6 +213,52 @@ func (ptgk *ptgBaseKinematics) ExecutionState(ctx context.Context) (motionplan.E
 	)
 }
 
+func deepCopyBaseState(original baseState) baseState {
```

Comment:
@biotinker 

This is how the data race is handled.
Since it is a write that is affecting a read in CheckPlan we create a deep copy of the currentState so that the values which are used in CheckPlan are not affected during course correction.

We need to do a deep copy, instead of a shallow copy to ensure that we have different addresses. With different addresses writing in course correction does not affect a CheckPlan that is happening in tandem.

Reply:
Can you write a test case for this?  Its hard for me to verify by inspection that its right.  If you issue a move call you should be able to get a `baseState` and then copy it.  Then the test would be if the copies line up the same

Reply:
While I agree this should negate the data race, it seems a bit excessive to make a copy of everything, and probably very computationally expensive; the trajectories can get fairly large, and to make additional duplicates with every call could impact our performance. Not to mention, this locking could actually significantly slow down course correction. It would be better to target specifically the offending piece of data.

The underlying culprit here appears to be that in the `courseCorrect` function, we are directly modifying fields in `connectionPoint`. When we initialize that with `connectionPoint := arcSteps[solution.stepIdx]`, if we instead make that a deep copy, then we will fix the problem while removing 99% of the data copying done here.

Reply:
we now do a deep copy of only the connectionPoint in execution.go. now there is no need for this function anymore.